{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyPmzQEemaBuWB0Mx8XPfM5C"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"pp5nOPARPJkB"},"outputs":[],"source":["import pandas as pd\n","import numpy as np\n","import random"]},{"cell_type":"code","source":["fileList = [\"12-30-2022sequences.txt\", \"12-30-salk.txt\", \"12-30not_clones.txt\", \"lotsOfSequences.txt\"]\n","\n","rightList = []\n","leftList = []\n","for thisFile in fileList:\n","\n","    #opens the file and turns it into a list of strings where\n","    #each item is a line in the file\n","    fhand = open(thisFile)\n","    flist = fhand.readlines()\n","    flist = [line.strip() for line in flist]\n","\n","    # makes list of positions where there is a \"COMMENT\" section in the file.\n","    # this is important becuase comment sections contain locus and position of sequence\n","    # relative to insertion (right or left border)\n","    linecount = 0\n","    commentLines = []\n","    for line in flist:\n","        if line[0:7] == \"COMMENT\":\n","            commentLines.append(linecount)\n","        linecount += 1\n","\n","    # goes to each comment section and extracts locus, right or left border position,\n","    # and the whole sequence itself\n","    seqList = []\n","    lcount = 0\n","    for com in commentLines:\n","        border = ''\n","        gene = ''\n","        commentrl = flist[com]\n","        commentgene = flist[com+1].split()\n","        gene = commentgene[-1]\n","        if \"right\" in commentrl:\n","            border = \"R\"\n","        if \"left\" in commentrl:\n","            border = \"L\"\n","            lcount += 1\n","        seqString = ''\n","        n = com\n","        for liney in flist[com:com+75]:\n","            n += 1\n","            if liney == \"ORIGIN\":\n","                while flist[n] != \"//\":\n","                    n += 1\n","                    seqString = seqString + flist[n]\n","        seqStringList = [i for i in seqString]\n","        newSeqList = []\n","        addSeq = True\n","        #ensures that sequences only contain a, g, c, t\n","        for char in seqStringList:\n","            if char == 'a' or char == 'c' or char == 'g' or char == 't':\n","                newSeqList.append(char)\n","            else:\n","              if char.isalpha():\n","                addSeq = False\n","                break\n","        if len(newSeqList) < 80:\n","          addSeq = False\n","        if addSeq:\n","            seqList.append([gene, border, newSeqList])\n","\n","    # reformats each sequence so that it includes the 80 closest bases to the insertion\n","    count = 0\n","    for i in seqList:\n","        if i[1] == 'L':\n","            i[2].append(\"1\")\n","            leftList.append(i[2][-81:])\n","        if i[1] == 'R':\n","            i[2][80] = 1\n","            rightList.append(i[2][:81])\n","\n","\n","\n","print(\"R:\", len(rightList))\n","print(\"L:\", len(leftList))\n","\n","#ensures that there are no duplicates in the final list\n","count = 0\n","for i in rightList:\n","    for n in rightList:\n","        if i == n:\n","            rightList.remove(i)\n","            count += 1\n","\n","count = 0\n","for i in leftList:\n","    for n in leftList:\n","        if i == n:\n","            leftList.remove(i)\n","            count += 1\n","\n","\n","print(\"R:\", len(rightList))\n","print(\"L:\", len(leftList))"],"metadata":{"id":"2A7gmUL8PWaQ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["dfR = pd.DataFrame(rightList)\n","dfR.head()\n","\n"],"metadata":{"id":"rrZzmq5doMoW"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#reads file containing whole Arabidopsis genome and converts it into a list of lists where\n","#each line in the file is an item in the fgenome list\n","fhandle = open('AT_genome_SEQUENCE.txt')\n","fgenome = fhandle.readlines()\n","fgenome = [line.strip() for line in fgenome]\n","fgenome = [[char.lower() for char in line] for line in fgenome]\n","\n","#ensures that empty lines in the file are not included in the fgenome list\n","emptyList = ['']\n","fgenome = [i for i in fgenome if i != emptyList]\n","\n","#adds \"0\" to the end of each item in fgenome\n","#serves as label for binary classification\n","for i in range(len(fgenome)):\n","    fgenome[i].append(\"0\")\n","\n","print(len(fgenome))"],"metadata":{"id":"s1fLftcwPf8G"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#import scipy.stats as stats\n","\n","\n","def GCcont(seq):\n","    ''' finds the percent GC content of a sequence as a decimal\n","\n","        seq: list of bases in sequence, in lowercase. assumes there is an extra\n","        character at the end denoting the label (in this case, 0 or 1)\n","    '''\n","    count = 0\n","    for base in seq:\n","        if base == 'g' or base == 'c':\n","            count += 1\n","    return (count/(len(seq)-1))\n","\n","def genNegList(borderList, fgenome):\n","  ''' generates a list of 'no insertion' sequences, AKA negative cases, such that\n","      the GC content distribution of the generated list is not statistically different\n","      from that of a specified list (borderList). Distributions are compared using\n","      a Mann-Whitney U test.\n","\n","      borderList: list of positive cases, generated list will have GC content distribution\n","                  similar to that of borderList. Additionally, generated list will be of\n","                  same length as borderList\n","\n","      fgenome: list of sequences that will be randomly sampled to generate list\n","               of negative cases\n","\n","  '''\n","\n","  GClist1 = []\n","  GClist0 = []\n","\n","  #generates list of GC contents for borderList to be compared to NegList later\n","  for item in borderList:\n","    GClist1.append(GCcont(item))\n","\n","\n","  pvalue = 0\n","  while pvalue < 0.4:\n","      pvalue = 0\n","      NegList = []\n","      GClist0 = []\n","      while len(NegList) < len(borderList):\n","          #randomly selects an index from fgenome\n","          x = random.randint(0, len(fgenome)-1)\n","          #ensures that the item of selected index from fgenome is a sequence\n","          if (fgenome[x][0] == 'g' or fgenome[x][0] == 'c' or fgenome[x][0] == 't' or fgenome[x][0] == 'a') and (len(fgenome[x]) == 81):\n","              #adds the GC content of selected item from fgenome to a copy of GClist0\n","              #and tests to ensure that distributions are not made statistically different\n","              #by adding this value\n","              GClist0copy = GClist0.copy()\n","              GClist0copy.append(GCcont(fgenome[x][0:79]))\n","              ustat, pvalue = stats.mannwhitneyu(GClist0copy, GClist1)\n","              #if added value does not make distributions significantly different,\n","              #it is added to NegList\n","              if pvalue > 0.4:\n","                  NegList.append(fgenome[x])\n","                  GClist0.append(GCcont(fgenome[x][0:79]))\n","      #p-value of final generated NegList is tested to ensure it is above 0.4\n","      ustat, pvalue = stats.mannwhitneyu(GClist0, GClist1)\n","      #prints p-value of final generated list\n","      print(pvalue)\n","\n","  #returns generated list of negative cases\n","  return NegList\n","\n","\n","#generates list of negative cases for right and left lists\n","rightNegList = genNegList(rightList, fgenome)\n","leftNegList = genNegList(leftList, fgenome)"],"metadata":{"id":"tYKZ56nqPi9i"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#combines positive and negative cases (insertion and no insertion) into one final list\n","CombinedRightList = rightList + rightNegList\n","CombinedLeftList = leftList + leftNegList\n","\n","\n","#ensures all items in left list are of appropriate length\n","line = 0\n","for i in CombinedLeftList:\n","    if len(i) != 81:\n","        CombinedLeftList.remove(i)\n","        print(\"(left list) wrong length on item\", line)\n","        line+=1\n","\n","#ensures that all items in right list are of appropriate length\n","line = 0\n","for i in CombinedRightList:\n","    if len(i) != 81:\n","        CombinedRightList.remove(i)\n","        print(\"(right list) wrong length on item\", line)\n","        line+=1\n","\n","#prints the length of each final list\n","print(\"R:\", len(CombinedRightList))\n","print(\"L:\", len(CombinedLeftList))\n","\n","#writes right list to .csv file\n","dfR = pd.DataFrame(CombinedRightList)\n","dfR.to_csv(input(\"file name: \"))\n","\n","#writes left list to .csv file\n","dfL = pd.DataFrame(CombinedLeftList)\n","dfL.to_csv(input(\"file name: \"))"],"metadata":{"id":"nqM57QxHPofA"},"execution_count":null,"outputs":[]}]}