{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyPmzQEemaBuWB0Mx8XPfM5C"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"pp5nOPARPJkB"},"outputs":[],"source":["import pandas as pd\n","import numpy as np\n","import random"]},{"cell_type":"code","source":["fileList = [\"12-30-2022sequences.txt\", \"12-30-salk.txt\", \"12-30not_clones.txt\", \"lotsOfSequences.txt\"]\n","\n","rightList = []\n","leftList = []\n","for thisFile in fileList:\n","\n","    #opens the file and turns it into a list of strings where\n","    #each item is a line in the file\n","    fhand = open(thisFile)\n","    flist = fhand.readlines()\n","    flist = [line.strip() for line in flist]\n","\n","    # makes list of positions where there is a \"COMMENT\" section in the file.\n","    # this is important becuase comment sections contain locus and position of sequence\n","    # relative to insertion (right or left border)\n","    linecount = 0\n","    commentLines = []\n","    for line in flist:\n","        if line[0:7] == \"COMMENT\":\n","            commentLines.append(linecount)\n","        linecount += 1\n","\n","    # goes to each comment section and extracts locus, right or left border position,\n","    # and the whole sequence itself\n","    seqList = []\n","    lcount = 0\n","    for com in commentLines:\n","        border = ''\n","        gene = ''\n","        commentrl = flist[com]\n","        commentgene = flist[com+1].split()\n","        gene = commentgene[-1]\n","        if \"right\" in commentrl:\n","            border = \"R\"\n","        if \"left\" in commentrl:\n","            border = \"L\"\n","            lcount += 1\n","        seqString = ''\n","        n = com\n","        for liney in flist[com:com+75]:\n","            n += 1\n","            if liney == \"ORIGIN\":\n","                while flist[n] != \"//\":\n","                    n += 1\n","                    seqString = seqString + flist[n]\n","        seqStringList = [i for i in seqString]\n","        newSeqList = []\n","        addSeq = True\n","        #ensures that sequences only contain a, g, c, t\n","        for char in seqStringList:\n","            if char == 'a' or char == 'c' or char == 'g' or char == 't':\n","                newSeqList.append(char)\n","            else:\n","              if char.isalpha():\n","                addSeq = False\n","                break\n","        if len(newSeqList) < 80:\n","          addSeq = False\n","        if addSeq:\n","            seqList.append([gene, border, newSeqList])\n","\n","    # reformats each sequence so that it includes the 80 closest bases to the insertion\n","    count = 0\n","    for i in seqList:\n","        if i[1] == 'L':\n","            i[2].append(\"1\")\n","            leftList.append(i[2][-81:])\n","        if i[1] == 'R':\n","            i[2][80] = 1\n","            rightList.append(i[2][:81])\n","\n","\n","\n","print(\"R:\", len(rightList))\n","print(\"L:\", len(leftList))\n","\n","#ensures that there are no duplicates in the final list\n","count = 0\n","for i in rightList:\n","    for n in rightList:\n","        if i == n:\n","            rightList.remove(i)\n","            count += 1\n","\n","count = 0\n","for i in leftList:\n","    for n in leftList:\n","        if i == n:\n","            leftList.remove(i)\n","            count += 1\n","\n","\n","print(\"R:\", len(rightList))\n","print(\"L:\", len(leftList))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"2A7gmUL8PWaQ","executionInfo":{"status":"ok","timestamp":1681659563898,"user_tz":300,"elapsed":614,"user":{"displayName":"Sawyer Smith","userId":"16467283874444835776"}},"outputId":"0fdb4c21-c0ef-4e56-815c-f2e0f14b0fad"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["R: 467\n","L: 726\n","R: 157\n","L: 244\n"]}]},{"cell_type":"code","source":["dfR = pd.DataFrame(rightList)\n","dfR.head()\n","\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":236},"id":"rrZzmq5doMoW","executionInfo":{"status":"ok","timestamp":1681659568188,"user_tz":300,"elapsed":221,"user":{"displayName":"Sawyer Smith","userId":"16467283874444835776"}},"outputId":"c2131142-09f6-462c-8bd7-1d4ce402e10c"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["  0  1  2  3  4  5  6  7  8  9   ... 71 72 73 74 75 76 77 78 79 80\n","0  g  g  g  g  a  t  c  a  g  t  ...  g  a  a  t  c  t  a  t  c  1\n","1  g  g  g  t  a  c  c  c  g  g  ...  t  g  a  t  c  a  t  a  t  1\n","2  a  t  c  t  g  g  a  a  a  t  ...  t  t  t  a  g  c  a  a  a  1\n","3  c  a  g  a  t  t  g  t  c  g  ...  a  a  g  g  t  t  t  t  c  1\n","4  a  g  g  g  t  a  c  c  c  g  ...  t  t  c  g  a  a  a  a  g  1\n","\n","[5 rows x 81 columns]"],"text/html":["\n","  <div id=\"df-3c61365b-733f-449e-83f8-25c81ea078d1\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>0</th>\n","      <th>1</th>\n","      <th>2</th>\n","      <th>3</th>\n","      <th>4</th>\n","      <th>5</th>\n","      <th>6</th>\n","      <th>7</th>\n","      <th>8</th>\n","      <th>9</th>\n","      <th>...</th>\n","      <th>71</th>\n","      <th>72</th>\n","      <th>73</th>\n","      <th>74</th>\n","      <th>75</th>\n","      <th>76</th>\n","      <th>77</th>\n","      <th>78</th>\n","      <th>79</th>\n","      <th>80</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>a</td>\n","      <td>t</td>\n","      <td>c</td>\n","      <td>a</td>\n","      <td>g</td>\n","      <td>t</td>\n","      <td>...</td>\n","      <td>g</td>\n","      <td>a</td>\n","      <td>a</td>\n","      <td>t</td>\n","      <td>c</td>\n","      <td>t</td>\n","      <td>a</td>\n","      <td>t</td>\n","      <td>c</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>t</td>\n","      <td>a</td>\n","      <td>c</td>\n","      <td>c</td>\n","      <td>c</td>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>...</td>\n","      <td>t</td>\n","      <td>g</td>\n","      <td>a</td>\n","      <td>t</td>\n","      <td>c</td>\n","      <td>a</td>\n","      <td>t</td>\n","      <td>a</td>\n","      <td>t</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>a</td>\n","      <td>t</td>\n","      <td>c</td>\n","      <td>t</td>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>a</td>\n","      <td>a</td>\n","      <td>a</td>\n","      <td>t</td>\n","      <td>...</td>\n","      <td>t</td>\n","      <td>t</td>\n","      <td>t</td>\n","      <td>a</td>\n","      <td>g</td>\n","      <td>c</td>\n","      <td>a</td>\n","      <td>a</td>\n","      <td>a</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>c</td>\n","      <td>a</td>\n","      <td>g</td>\n","      <td>a</td>\n","      <td>t</td>\n","      <td>t</td>\n","      <td>g</td>\n","      <td>t</td>\n","      <td>c</td>\n","      <td>g</td>\n","      <td>...</td>\n","      <td>a</td>\n","      <td>a</td>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>t</td>\n","      <td>t</td>\n","      <td>t</td>\n","      <td>t</td>\n","      <td>c</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>a</td>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>g</td>\n","      <td>t</td>\n","      <td>a</td>\n","      <td>c</td>\n","      <td>c</td>\n","      <td>c</td>\n","      <td>g</td>\n","      <td>...</td>\n","      <td>t</td>\n","      <td>t</td>\n","      <td>c</td>\n","      <td>g</td>\n","      <td>a</td>\n","      <td>a</td>\n","      <td>a</td>\n","      <td>a</td>\n","      <td>g</td>\n","      <td>1</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows Ã— 81 columns</p>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3c61365b-733f-449e-83f8-25c81ea078d1')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-3c61365b-733f-449e-83f8-25c81ea078d1 button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-3c61365b-733f-449e-83f8-25c81ea078d1');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":9}]},{"cell_type":"code","source":["#reads file containing whole Arabidopsis genome and converts it into a list of lists where\n","#each line in the file is an item in the fgenome list\n","fhandle = open('AT_genome_SEQUENCE.txt')\n","fgenome = fhandle.readlines()\n","fgenome = [line.strip() for line in fgenome]\n","fgenome = [[char.lower() for char in line] for line in fgenome]\n","\n","#ensures that empty lines in the file are not included in the fgenome list\n","emptyList = ['']\n","fgenome = [i for i in fgenome if i != emptyList]\n","\n","#adds \"0\" to the end of each item in fgenome\n","#serves as label for binary classification\n","for i in range(len(fgenome)):\n","    fgenome[i].append(\"0\")\n","\n","print(len(fgenome))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":235},"id":"s1fLftcwPf8G","executionInfo":{"status":"error","timestamp":1683125966653,"user_tz":300,"elapsed":1295,"user":{"displayName":"Sawyer Smith","userId":"16467283874444835776"}},"outputId":"b4ca7574-eeca-45e6-8bcc-11f2c055234b"},"execution_count":null,"outputs":[{"output_type":"error","ename":"FileNotFoundError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-2-7d54918f1c52>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#reads file containing whole Arabidopsis genome and converts it into a list where\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m#each line in the file is an item in the list\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mfhandle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'AT_genome_SEQUENCE.txt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mfgenome\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfhandle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mfgenome\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfgenome\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'AT_genome_SEQUENCE.txt'"]}]},{"cell_type":"code","source":["#import scipy.stats as stats\n","\n","\n","def GCcont(seq):\n","    ''' finds the percent GC content of a sequence as a decimal\n","\n","        seq: list of bases in sequence, in lowercase. assumes there is an extra\n","        character at the end denoting the label (in this case, 0 or 1)\n","    '''\n","    count = 0\n","    for base in seq:\n","        if base == 'g' or base == 'c':\n","            count += 1\n","    return (count/(len(seq)-1))\n","\n","def genNegList(borderList, fgenome):\n","  ''' generates a list of 'no insertion' sequences, AKA negative cases, such that\n","      the GC content distribution of the generated list is not statistically different\n","      from that of a specified list (borderList). Distributions are compared using\n","      a Mann-Whitney U test.\n","\n","      borderList: list of positive cases, generated list will have GC content distribution\n","                  similar to that of borderList. Additionally, generated list will be of\n","                  same length as borderList\n","\n","      fgenome: list of sequences that will be randomly sampled to generate list\n","               of negative cases\n","\n","  '''\n","\n","  GClist1 = []\n","  GClist0 = []\n","\n","  #generates list of GC contents for borderList to be compared to NegList later\n","  for item in borderList:\n","    GClist1.append(GCcont(item))\n","\n","\n","  pvalue = 0\n","  while pvalue < 0.4:\n","      pvalue = 0\n","      NegList = []\n","      GClist0 = []\n","      while len(NegList) < len(borderList):\n","          #randomly selects an index from fgenome\n","          x = random.randint(0, len(fgenome)-1)\n","          #ensures that the item of selected index from fgenome is a sequence\n","          if (fgenome[x][0] == 'g' or fgenome[x][0] == 'c' or fgenome[x][0] == 't' or fgenome[x][0] == 'a') and (len(fgenome[x]) == 81):\n","              #adds the GC content of selected item from fgenome to a copy of GClist0\n","              #and tests to ensure that distributions are not made statistically different\n","              #by adding this value\n","              GClist0copy = GClist0.copy()\n","              GClist0copy.append(GCcont(fgenome[x][0:79]))\n","              ustat, pvalue = stats.mannwhitneyu(GClist0copy, GClist1)\n","              #if added value does not make distributions significantly different,\n","              #it is added to NegList\n","              if pvalue > 0.4:\n","                  NegList.append(fgenome[x])\n","                  GClist0.append(GCcont(fgenome[x][0:79]))\n","      #p-value of final generated NegList is tested to ensure it is above 0.4\n","      ustat, pvalue = stats.mannwhitneyu(GClist0, GClist1)\n","      #prints p-value of final generated list\n","      print(pvalue)\n","\n","  #returns generated list of negative cases\n","  return NegList\n","\n","\n","#generates list of negative cases for right and left lists\n","rightNegList = genNegList(rightList, fgenome)\n","leftNegList = genNegList(leftList, fgenome)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"tYKZ56nqPi9i","executionInfo":{"status":"ok","timestamp":1681659579386,"user_tz":300,"elapsed":920,"user":{"displayName":"Sawyer Smith","userId":"16467283874444835776"}},"outputId":"55c32601-b377-4eff-9176-3d2f8c08cf4e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["0.40880884683881025\n","0.40685765417495623\n"]}]},{"cell_type":"code","source":["#combines positive and negative cases (insertion and no insertion) into one final list\n","CombinedRightList = rightList + rightNegList\n","CombinedLeftList = leftList + leftNegList\n","\n","\n","#ensures all items in left list are of appropriate length\n","line = 0\n","for i in CombinedLeftList:\n","    if len(i) != 81:\n","        CombinedLeftList.remove(i)\n","        print(\"(left list) wrong length on item\", line)\n","        line+=1\n","\n","#ensures that all items in right list are of appropriate length\n","line = 0\n","for i in CombinedRightList:\n","    if len(i) != 81:\n","        CombinedRightList.remove(i)\n","        print(\"(right list) wrong length on item\", line)\n","        line+=1\n","\n","#prints the length of each final list\n","print(\"R:\", len(CombinedRightList))\n","print(\"L:\", len(CombinedLeftList))\n","\n","#writes right list to .csv file\n","dfR = pd.DataFrame(CombinedRightList)\n","dfR.to_csv(input(\"file name: \"))\n","\n","#writes left list to .csv file\n","dfL = pd.DataFrame(CombinedLeftList)\n","dfL.to_csv(input(\"file name: \"))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"nqM57QxHPofA","executionInfo":{"status":"ok","timestamp":1681659607729,"user_tz":300,"elapsed":23039,"user":{"displayName":"Sawyer Smith","userId":"16467283874444835776"}},"outputId":"855daf9a-55c9-48b8-c97a-1a8afff53388"},"execution_count":null,"outputs":[{"name":"stdout","output_type":"stream","text":["R: 314\n","L: 488\n","file name: agroRight.csv\n","file name: agroLeft.csv\n"]}]}]}